### DISCUSSION TITLE
Confusion with ngrok headers

### DISCUSSION THREAD

**Discussion by lnpbk** [laravel/valet#1280] [Q&A]
Hi,

Valet share recently started failing for me when using Stripe webhooks. I'm running this command:

```
valet share app.xxxxxx --region=eu
```

I found that in `server.php` that the value for `HTTP_X_FORWARDED_HOST` is being set to `HTTP_X_ORIGINAL_HOST` from ngrok:

```
/**
 * ngrok uses the X-Original-Host to store the forwarded hostname.
 */
if (isset($_SERVER['HTTP_X_ORIGINAL_HOST']) && ! isset($_SERVER['HTTP_X_FORWARDED_HOST'])) {
    $_SERVER['HTTP_X_FORWARDED_HOST'] = $_SERVER['HTTP_X_ORIGINAL_HOST'];
}
```

In my use-case, this appears to be the opposite of what is desired.

Requests through ngrok set the `X-Original-Host` header to the ngrok url, and the `Host` header is set to my local valet domain.

The `frontControllerPath` method in `LaravelValetDriver` then sets the `HTTP_HOST` from this `HTTP_X_FORWARDED_HOST` value:

```
// Shortcut for getting the "local" hostname as the HTTP_HOST
if (isset($_SERVER['HTTP_X_ORIGINAL_HOST'], $_SERVER['HTTP_X_FORWARDED_HOST'])) {
    $_SERVER['HTTP_HOST'] = $_SERVER['HTTP_X_FORWARDED_HOST'];
}
```

This means that `HTTP_X_FORWARDED_HOST`, `HTTP_X_ORIGINAL_HOST`, and `HTTP_HOST` all get set to the ngrok url, and my routes with domain restriction fail with 404's.

If I comment out the lines in `server.php`, everything works as expected. Am I missing something here?

Thanks!

**Comment by mattstauffer** [comment]
I found the original discussion about why this works this way:

https://github.com/laravel/valet/issues/342

Adam writes:

> If I'm not mistaken, it was an issue with how URLs would get generated when using Laravel helpers like url() when sharing a site with ngrok. They would end up pointing at the wrong place if we didn't override that header.
> 
> If I had a site setup locally as my-site.dev, and then I shared that and got some ngrok URL, as soon as someone tried to click a link in the ngrok version of the site, it would try to take them to my-site.dev which of course wouldn't work for them.

I'm sadly not sure how to modify this code to solve for your concern (Laravel correctly knowing the local site you're trying to represent) and theirs (Laravel correctly generating links to the ngrok version of URLs, not the local version).
